velox/functions/prestosql/ArrayPosition.cpp

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <deque>
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/vector/DecodedVector.h"
+
+namespace facebook::velox::functions {
+namespace {
+
+template <TypeKind kind>
+void applyTypedFirstMatch(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
```

Comment:
Any particular reason elementsDecoded is const, but array and searchDecoded are not?

Reply:
Good questions. `arrayDecoded` calls `DecodedVector::indices()` inside the function body, and `DecodedVector::indices()` is not a const method, so `arrayDecoded` cannot be const. The same thing applies to `searchDecoded` that calls `indices()` inside `applyComplexTypeFirstMatch()` (and a specialization of `applyTypedFirstMatch()` calls `applyComplexTypeFirstMatch()`, so `searchDecoded` cannot be const in both functions).